### PR TITLE
[FROZEN] Update withdraw state based on blockchain

### PR DESCRIPTION
**Pull request by ysv**

### CODE REVIEW

**Comment by ysv** [comment]
@yivo take a look please

File: app/models/withdraw.rb [hunk 1]
```diff
@@ -133,25 +138,26 @@ def send_coins!
...
+#  txid          :string(128)
+#  aasm_state    :string(30)       not null
+#  sum           :decimal(32, 16)  not null
+#  type          :string(30)       not null
+#  tid           :string(64)       not null
+#  rid           :string(64)       not null
+#  created_at    :datetime         not null
+#  updated_at    :datetime         not null
+#  completed_at  :datetime
+#  confirmations :integer          default(0), not null
```
Comment (yivo) [inline app/models/withdraw.rb]: Should be 4 byte integer unsigned not null default 0 and located after rid.

File: lib/daemons/withdraw_confirm.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+
+require File.join(ENV.fetch('RAILS_ROOT'), 'config', 'environment')
+
+running = true
+Signal.trap(:TERM) { running = false }
+
+while running
+  Withdraw.confirming.order(created_at: :asc).map(&:try_to_confirm!)
```
Comment (yivo) [inline lib/daemons/withdraw_confirm.rb]: Never use created_at for ordering. It is not in the index list. Use `Withdraw::Coin.includes(:currency).find_each(bulk_size: 10) { |record| ... }`. Check if option `bulk_size` is correct, I am not sure.

You need to handle withdraw_confirmation = 0, first load confirmations for TXID, then reload record by acquiring lock on it and atomically update state.

File: app/models/withdraws/coin.rb [hunk 3]
```diff
@@ -43,6 +43,17 @@ def audit!
       end
     end
 
+    def try_to_confirm!
```
Comment (yivo) [inline app/models/withdraws/coin.rb]: @yssavchuk Delete this method and move it to daemon.

File: lib/daemons/withdraw_confirm.rb (hunk 2 above)
Comment (yivo) [reply lib/daemons/withdraw_confirm.rb]: You also missed break unless running inside loop.

**Review by yivo — CHANGES_REQUESTED** [review]

**Comment by ysv** [comment]
@yivo updated

File: app/models/withdraw.rb [hunk 4]
```diff
@@ -65,9 +66,13 @@ class Withdraw < ActiveRecord::Base
       after :send_coins!
     end
 
+    event :confirm do
```
Comment (yivo) [inline app/models/withdraw.rb]: Well it is not logical name for the state which means not finished operation. `withdraw.confirm` seems like it was confirmed in blockchain. Change name to something like `dispatch`.

File: lib/daemons/withdraw_confirm.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+# frozen_string_literal: true
+
+require File.join(ENV.fetch('RAILS_ROOT'), 'config', 'environment')
+
+running = true
+Signal.trap(:TERM) { running = false }
+
+while running
+  Withdraw::Coin.confirming.includes(:currency).find_each(batch_size: 10) do |withdraw|
+    break if withdraw.txid.blank?
```
Comment (yivo) [inline lib/daemons/withdraw_confirm.rb]: Why it should be blank here? Remove the like pls.

**Review by yivo — CHANGES_REQUESTED** [review]

**Review by yivo — APPROVED** [review]

File: lib/coin_api/bit_go.rb [hunk 6]
```diff
@@ -115,13 +115,14 @@ def escape_path_component(id)
...
+                  confirmations: tx.fetch('confirmations').to_i,
+                  received_at:   Time.parse(tx.fetch('date')) }
+      return deposit unless with_entries
       entries = build_deposit_entries(tx)
       return if entries.blank?
-      { id:            normalize_txid(tx.fetch('id')),
-        confirmations: tx.fetch('confirmations').to_i,
-        entries:       entries,
-        received_at:   Time.parse(tx.fetch('date')) }
+      deposit.merge(entries: entries)
```
Comment (ysv) [inline lib/coin_api/bit_go.rb:125]: @yivo I found this solution. What do you think ?

**Comment by ysv** [comment]
Closed in #1427

**Pull request closed without merging**